lib/codecs/src/encoding/format/syslog.rs

```diff
@@ -0,0 +1,577 @@
+    /// Proc ID
+    proc_id: Option<String>,
+    /// Msg ID
+    msg_id: Option<String>,
+
+    /// Payload key
+    payload_key: String,
+
+    // Q: The majority of the fields above pragmatically only make sense as config for keys to query?
+}
+
+/// Serializer that converts an `Event` to bytes using the Syslog format.
+#[derive(Debug, Clone)]
+pub struct SyslogSerializer {
+    config: SyslogSerializerConfig
+}
+
+impl SyslogSerializer {
+    /// Creates a new `SyslogSerializer`.
+    pub fn new(conf: &SyslogSerializerConfig) -> Self {
+        Self { config: conf.clone() }
+    }
+}
+
+impl Encoder<Event> for SyslogSerializer {
+    type Error = vector_common::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        if let Event::Log(log_event) = event {
+            let syslog_message = ConfigDecanter::new(log_event).decant_config(&self.config.syslog);
+
+            let vec = syslog_message
+                .encode(&self.config.syslog.rfc)
+                .as_bytes()
+                .to_vec();
+            buffer.put_slice(&vec);
+        }
+
+        Ok(())
+    }
+}
+
+// Adapts a `LogEvent` into a `SyslogMessage` based on config from `SyslogSerializerOptions`:
+// - Splits off the responsibility of encoding logic to `SyslogMessage` (which is not dependent upon Vector types).
+// - Majority of methods are only needed to support the `decant_config()` operation.
+struct ConfigDecanter {
+    log: LogEvent,
+}
+
+impl ConfigDecanter {
+    fn new(log: LogEvent) -> Self {
+        Self {
+            log,
+        }
+    }
+
+    fn decant_config(&self, config: &SyslogSerializerOptions) -> SyslogMessage {
+        let y = |v| self.replace_if_proxied_opt(v);
+        let app_name = y(&config.app_name).unwrap_or("vector".to_owned());
+        let proc_id = y(&config.proc_id);
+        let msg_id = y(&config.msg_id);
+
+        SyslogMessage {
+            pri: Pri {
+                facility: config.facility,
+                severity: config.severity,
+            },
+            timestamp: self.get_timestamp(),
+            hostname: self.value_by_key("hostname"),
+            tag: Tag {
+                app_name,
+                proc_id,
+                msg_id,
+            },
+            structured_data: self.get_structured_data(),
+            message: self.get_message(&config),
+        }
+    }
+
+    fn replace_if_proxied_opt(&self, value: &Option<String>) -> Option<String> {
+        value.as_ref().and_then(|v| self.replace_if_proxied(v))
+    }
+
+    // When the value has the expected prefix, perform a lookup for a field key without that prefix part.
+    // A failed lookup returns `None`, while a value without the prefix uses the config value as-is.
+    //
+    // Q: Why `$.message.` as the prefix? (Appears to be JSONPath syntax?)
+    // NOTE: Originally named in PR as: `get_field_or_config()`
+    fn replace_if_proxied(&self, value: &str) -> Option<String> {
```

Comment:
I don't understand what this is doing, but whatever it is doing it is not what we want it to be doing... These function can just be removed given that the fields should be retrieved with something like

```
field = value.unwrap_or_else(|| log.get_field_by_meaning("service"))
```

With a different meaning for each field...

Reply:
The original PR IIRC had implemented a feature to support config that overrides a default value (such as always X severity), presumably because the log event did not provide an equivalent mapping. Or the config key would reference a field to lookup and get the value of from the event, which I think their `$.message` prefix was specific to their own use-case.

I'd rather drop it, but wasn't sure how much of the original PR was the correct approach as I'm not familiar enough with Vector myself. Thus this PR focused on refactoring the functionality that was already in the existing PR and moving forward from there 😅 

---

So this method was mostly a helper:

```rust
        value
            .strip_prefix("$.message.")
            .map_or(
                Some(value.to_owned()),
                |field_key| self.value_by_key(field_key),
            )
    }

    fn value_by_key(&self, field_key: &str) -> Option<String> {
        self.log.get(field_key).map(|field_value| {
            field_value.to_string_lossy().to_string()
        })
    }
```

Used earlier like:

```rust
    fn decant_config(&self, config: &SyslogSerializerOptions) -> SyslogMessage {
        let y = |v| self.replace_if_proxied_opt(v);
        let app_name = y(&config.app_name).unwrap_or("vector".to_owned());
        let proc_id = y(&config.proc_id);
        let msg_id = y(&config.msg_id);
```

Since they're options, if no lookup key or default value was configured, `y` would return `None`:
- `proc_id` and `msg_id` would later be formatted as `-`.
- `app_name` default is meant to fallback to something IIRC so `vector` was the default chosen. In my local branch I used the `SmartDefault` crate to add a default to the actual syslog `Tag` struct for the `app_name` field instead.

---

I'm not familiar with `log.get_field_by_meaning` or if that was available when I started working on this PR early this year. I'll look into that if it's appropriate (_I'm aware of [this reference](https://github.com/vectordotdev/vector/pull/17668#discussion_r1761846388) you shared in another feedback comment, I haven't gone over that yet_).

IIRC this logic may have been refactored on my local branch.

Reply:
> The original PR IIRC had implemented a feature to support config that overrides a default value (such as always X severity), presumably because the log event did not provide an equivalent mapping. Or the config key would reference a field to lookup and get the value of from the event, which I think their $.message prefix was specific to their own use-case.

This came out of translating functionality from an existing fluentd plugin.  One can accomplish similar need by using existing vector functionality to template config field and walk a path.  IMO you can remove this functionality if it is possible to do a similar way in vector already

Reply:
> IMO you can remove this functionality if it is possible to do a similar way in vector already

Pretty sure I used VRL with a remap transform in my vector config to get similar functionality. I'm not fond of using config settings for the encoder to support individual fields that could be either a fixed value or lookup key.

When I've had time to refresh my familiarity with the PR source and sync my local changes to this branch we can discuss further.

Either way it'll need to clearly document how a user can map their source log event into the expected input/format that the encoder needs, relying on config settings for this was mostly a convenience IIRC but introduced some added complexity that I don't think was common with other official Vector encoders.